pkg/controller/constraint/constraint_controller.go

```diff
@@ -274,7 +343,83 @@ func (r *ReconcileConstraint) Reconcile(ctx context.Context, request reconcile.R
 		status.Status.ConstraintUID = instance.GetUID()
 		status.Status.ObservedGeneration = instance.GetGeneration()
 		status.Status.Errors = nil
-		if c, err := r.cfClient.GetConstraint(instance); err != nil || !constraints.SemanticEqual(instance, c) {
+
+		if c, err := r.cfClient.GetConstraint(instance); err != nil || !reflect.DeepEqual(instance, c) {
+			// generate vapbinding resources
+			if generateVapBinding && IsVapAPIEnabled() {
+				currentVapBinding := &admissionregistrationv1beta1.ValidatingAdmissionPolicyBinding{}
+				vapBindingName := fmt.Sprintf("gatekeeper-%s", instance.GetName())
+				log.Info("check if vapbinding exists", "vapBindingName", vapBindingName)
+				if err := r.reader.Get(ctx, types.NamespacedName{Name: vapBindingName}, currentVapBinding); err != nil {
+					if !apierrors.IsNotFound(err) && !errors.As(err, &discoveryErr) && !meta.IsNoMatchError(err) {
+						return reconcile.Result{}, err
+					}
+					currentVapBinding = nil
+				}
+				newVapBinding := &admissionregistrationv1beta1.ValidatingAdmissionPolicyBinding{}
+				transformedVapBinding, err := transform.ConstraintToBinding(instance)
+				if err != nil {
+					status.Status.Errors = append(status.Status.Errors, constraintstatusv1beta1.Error{Message: err.Error()})
+					if err2 := r.writer.Update(ctx, status); err2 != nil {
+						log.Error(err2, "could not report transform vapbinding error status")
+					}
+					return reconcile.Result{}, err
+				}
+				if currentVapBinding == nil {
+					newVapBinding = transformedVapBinding.DeepCopy()
+				} else {
+					newVapBinding = currentVapBinding.DeepCopy()
+					newVapBinding.Spec = transformedVapBinding.Spec
+				}
+
+				if err := controllerutil.SetControllerReference(instance, newVapBinding, r.scheme); err != nil {
+					return reconcile.Result{}, err
+				}
+
+				if currentVapBinding == nil {
+					log.Info("creating vapbinding")
+					if err := r.writer.Create(ctx, newVapBinding); err != nil {
+						status.Status.Errors = append(status.Status.Errors, constraintstatusv1beta1.Error{Message: err.Error()})
+						if err2 := r.writer.Update(ctx, status); err2 != nil {
+							log.Error(err2, "could not report creating vapbinding error status")
+						}
+						return reconcile.Result{}, err
+					}
+				}
+				if !reflect.DeepEqual(currentVapBinding, newVapBinding) {
```

Comment:
nit: I'd make this an else if statement, otherwise we make two writes to the API server instead of one (on account of currentVAP being nil, guaranteeing reflect.DeepEqual() will be false.